### PR TITLE
deps: update ibis to version 8.0.0 and refactor `remote_function` to use ibis UDF method

### PR DESCRIPTION
**Pull request by tswast**
Thank you for opening a Pull Request! Before submitting your PR, there are a few things you can do to make sure it goes smoothly:
- [ ] Make sure to open an issue as a [bug/issue](https://github.com/googleapis/python-bigquery-dataframes/issues/new/choose) before writing your code!  That way we can discuss the change, evaluate designs, and agree on the general idea
- [ ] Ensure the tests and linter pass
- [ ] Code coverage does not decrease (if any source code was changed)
- [ ] Appropriate docs were updated (if necessary)

Follow-up to https://github.com/googleapis/python-bigquery-dataframes/pull/53#discussion_r1427224630
🦕

### CODE REVIEW

**Comment by tswast** [comment]
I'm trying to debug the remaining unit test. It seems to be the string version of row_number() for order ID is the one that is ending up as an unresolvable operation.

Might be this reason?

```python
import ibis
import pandas

df = pandas.DataFrame(
            {
                "col1": [1, 2, 3],
                "col2": ["a", "b", "c"],
                "col3": [0.1, 0.2, 0.3],
            }
        )
client = ibis.pandas.connect({"test_table": df})
table = client.table("test_table")
order_window = ibis.window(order_by=[ibis.asc("col1")])
order_id = ibis.row_number().over(order_window)
order_id
table[order_id]
table.select(table["col1"], order_id)
client.execute(table.select(table["col1"], order_id))
```

```
File ~/envs/bigframes/lib/python3.10/site-packages/ibis/backends/pandas/dispatch.py:26, in raise_unknown_op(node, *args, **kwargs)
     23 @execute_node.register(ops.Node, [object])
     24 def raise_unknown_op(node, *args, **kwargs):
     25     signature = ", ".join(type(arg).__name__ for arg in args)
---> 26     raise com.OperationNotDefinedError(
     27         "Operation is not implemented for this backend with "
     28         f"signature: execute_node({type(node).__name__}, {signature})"
     29     )

OperationNotDefinedError: Operation is not implemented for this backend with signature: execute_node(RowNumber, )
```

**Comment by tswast** [comment]
Looks like the regression in `nsmallest` tests are due to https://github.com/ibis-project/ibis/issues/7940

**Comment by tswast** [comment]
Looks like there are some changes I need to make to the remote functions type inference to account for the ibis 8.0.0 upgrade.

The `nlargest` / `nsmallest` tests are still failing though. I think I need to bump the minimum sqlglot: https://github.com/tobymao/sqlglot/blob/main/CHANGELOG.md#v2080---2024-01-08

File: third_party/bigframes_vendored/ibis/backends/bigquery/datatypes.py [hunk 1]
```diff
@@ -0,0 +1,176 @@
+# Contains code from
```
Comment (chelsea-lin) [inline third_party/bigframes_vendored/ibis/backends/bigquery/datatypes.py:1]: curious why we need a copy of these codes. Any customization for BigFrames or any known issues in ibis to be worked around?

**Review by chelsea-lin — APPROVED** [review]

File: third_party/bigframes_vendored/ibis/backends/bigquery/datatypes.py (hunk 1 above)
Comment (tswast) [reply third_party/bigframes_vendored/ibis/backends/bigquery/datatypes.py:1]: Ibis now maps to the sqlglot type aliases, but we still needed the final BigQuery type for some remote functions code.

**Pull request merged**